scripts/usb_scan.py

```diff
@@ -0,0 +1,545 @@
+        Log().debug("policy list: " + str(self.rule_list))
+
+    def parse_line(self, line):
+        # 1. remove comments
+        # ^([^#]*)(#.*)?$
+        i = line.find("#")
+        if i != -1:
+            line = line[0:i]
+
+        # 2. split action and match field
+        # ^\s*(ALLOW|DENY)\s*:\s*([^:]*)$
+        if ":" not in line:
+            return
+        parts = line.split(":")
+        n = len(parts)
+        if n == 0 or n > 2:
+            return
+
+        # 3. parse action
+        # \s*(ALLOW|DENY)\s*
+        rule = {}
+        action = parts[0].strip()
+        if action == "ALLOW":
+            rule[self._ALLOW] = True
+        elif action == "DENY":
+            rule[self._ALLOW] = False
+        else:
+            return
+
+        # 4. parse key=value pairs
+        # pattern = r"\s*(class|subclass|prot|vid|pid|rel)\s*=\s*([0-9a-fA-F]+|\"[0-9a-fA-F]+\")"
+        if n == 2:
+            beg, target = 0, parts[1].strip()
+            for matchNum, match in enumerate(re.finditer(self._PATTERN, target)):
+                if beg != match.start():
+                    Log().debug("policy line ignored: " + line)
+                    return
+                beg = match.end()
+                if match.groups()[0] not in self.rule_list:
+                    rule[match.groups()[0]] = match.groups()[1].strip('"')
+            if beg != len(target):
+                Log().debug("policy line ignored: " + line)
+                return
+
+        self.rule_list.append(rule)
+
+    @staticmethod
+    def hex_equal(h1, h2):
+        # Log().debug("h1: " + str(h1) + "  h2: " + h2)
+        try:
+            return int(h1, 16) == int(h2, 16)
+        except ValueError:
+            return False
+
+    def match_device_interface(self, rule, device, interface):
+        for k in [k for k in rule if k in self._KEY_MAP_DEVICE]:
+            Log().debug("check {} props[{}] against {}".format(interface.get_node(), k, str(rule)))
+            if not self.hex_equal(rule[k], device[self._KEY_MAP_DEVICE[k]]):
+                return False
+
+        for k in [k for k in rule if k in self._KEY_MAP_INTERFACE]:
+            Log().debug("check {} props[{}] against {}".format(interface.get_node(), k, str(rule)))
+            if not self.hex_equal(rule[k], interface[self._KEY_MAP_INTERFACE[k]]):
+                return False
+
+        Log().debug("found matching rule: " + str(rule))
+        return True
+
+    def check(self, device):
+        Log().debug("policy check: " + device.get_node())
+        allow = False
+        for rule in self.rule_list:
+            for interface in device.get_all_interfaces():
+                if self.match_device_interface(rule, device, interface):
+                    if not rule[self._ALLOW]:
+                        Log().debug("deny " + device.get_node())
+                        return False
+                    else:
+                        allow = True
+                        continue
```

Comment:
Surely this just causes the inner loop to repeat, which was going to happen anyway?

Reply:
TODO, I need to skip checking this interface if allow.
continue is not enough,
I should remove it.